runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -387,10 +424,15 @@ public boolean isWriteLocked() {
             }
         }
         // Optimistic reading failed, retry with a full lock
-        ts = lock.readLock();
+        try {
+            ts = lock.tryReadLock(1, TimeUnit.SECONDS);
+        } catch (InterruptedException ie) {
+            log.debug("ReadLock[{}] Timed out, retrying with write lock");
+            throw new ConcurrentModificationException();
+        }
         try {
             try {
-                return readFunction.apply(version, object);
+                return readFunction.apply(getVersionUnsafe(), this);
             } finally {
                 lock.unlock(ts);
```

Comment:
Michael: when using Roger's MultipleNonOverlappingTest test case + adding some artificial delays in the runtime + commit f1e0dfcdc67fc09f075f55aeef07580c63dc0270 ("reduce context calls"), I see this unlock call throwing java.lang.IllegalMonitorStateException.